waspc/cli/Command/Watch.hs

```diff
@@ -35,19 +44,45 @@ watch waspProjectDir outDir = FSN.withManager $ \mgr -> do
   chan <- newChan
   _ <- FSN.watchDirChan mgr (SP.toFilePath waspProjectDir) eventFilter chan
   _ <- FSN.watchTreeChan mgr (SP.toFilePath $ waspProjectDir </> Common.extCodeDirInWaspProjectDir) eventFilter chan
-  listenForEvents chan currentTime
+  timeOfLastEvent <- MVar.newMVar currentTime
+  listenForEvents chan currentTime timeOfLastEvent
   where
-    listenForEvents :: Chan FSN.Event -> UTCTime -> IO ()
-    listenForEvents chan lastCompileTime = do
+    oneSecond :: Int
+    oneSecond = 1000000
+
+    oneSecondDelay :: MVar.MVar UTCTime -> IO ()
+    oneSecondDelay timeOfLastEventMVar = do
+      threadDelay oneSecond
+      currentDelayTime <- MVar.readMVar timeOfLastEventMVar
+      currentTime <- getCurrentTime
+      let timeDiff = nominalDiffTimeToSeconds $ diffUTCTime currentTime currentDelayTime
+      when (timeDiff < fromIntegral oneSecond) $ do
+        threadDelay (oneSecond - (floor . (* 1e9) $ timeDiff))
+        oneSecondDelay timeOfLastEventMVar
+
+    recurEventCheck :: Chan FSN.Event -> UTCTime -> MVar.MVar UTCTime -> IO ()
+    recurEventCheck chan currentTime timeOfLastEventMVar = do
+      event <- readChan chan
+      let eventTime = FSN.eventTime event
+      unless (eventTime < currentTime) $ do
```

Comment:
I think what you want to do here is ensure that MVar is set to the time of the latest event. I would do that by reading the value of MVar, taking maximum of eventTime and MVar, and then would put that into MVar. That why you always have the time of the latest event in MVar.

You don't need to pass `currentTime` for this to the recurEventCheck, so you can drop that argument.

However, you will need to call recurEventCheck again at the end of recurEventCheck -> otherwise there is no recursion and no recurrence! Right now, as you implemented it, recurEventCheck executes only once and that is it, so it will read only one event and end.